job_runner.py

```diff
@@ -0,0 +1,97 @@
+
+        print("\n\n\n")
+        print_yellow("{:#^90}".format("   Compliance Testing Summary   "))
+        print_yellow("#{:^88}#".format(""))
+        print_yellow("#{:^88}#".format(f"Total Tests - {self.test_count}"))
+        print_yellow("#{:^88}#".format(f'Passed - {self.result["passed"]} ({passed_tests})'))
+        print_yellow("#{:^88}#".format(f'Failed - {self.result["failed"]} ({failed_tests})'))
+        print_yellow("#{:^88}#".format(f'Skipped - {self.result["skipped"]} ({skipped_tests})'))
+        print_yellow("#{:^88}#".format(""))
+        print_yellow("{:#^90}".format(""))
+
+    def validate_job(self, yaml_data, yaml_file):
+        schema_path = os.path.join(self.path, "tests", "template", "test_template_schema.json")
+        with open(schema_path, "r") as f:
+            json_schema = yaml.safe_load(f)
+
+        try:
+            validate(yaml_data, json_schema)
+            print(f'Test YAML file valid for {yaml_file}')
+        except ValidationError as err:
+            raise ComplianceException(f"YAML schema validation error - {yaml_file}. Error details - {err.message}")
+
+    def validate_yaml(self, file, yaml_file):
+        try:
+            yaml_data = yaml.safe_load(file)
+            return yaml_data
+        except yaml.YAMLError as err:
+            raise ComplianceException(f"Invalid YAML File - {yaml_file}. Error details - {err}")
+
+    def run_jobs(self):
+
+        yaml_path = os.path.join(self.path, "tests")
+        for yaml_file in os.listdir(yaml_path):
+            if yaml_file.endswith(".yml"):
+                self.test_count += 1
+                with open(os.path.join(yaml_path, yaml_file), "r") as f:
+                    print("\n{:#^100}".format(f"     Initiating Test-{self.test_count} for {yaml_file}     "))
+                    try:
+                        yaml_data = self.validate_yaml(f, yaml_file)
+                        self.validate_job(yaml_data, yaml_file)
+
+                        tag_matched = False
+                        # print(f' {self.tags}   {yaml_data["tags"]}')
+                        for tag in self.tags:
+                            if tag in yaml_data["tags"]:
+                                tag_matched = True
+                                test_runner = TestRunner(yaml_data["server"], yaml_data["version"][0])
+                                job_count = 0
+                                for job in yaml_data["jobs"]:
+                                    job_count += 1
+                                    print(f'Running tests for sub-job-{job_count} -> {job["name"]}')
+                                    test_runner.run_tests(job)
+                                self.result["passed"] += 1
+                                self.test_status["passed"].append(str(self.test_count))
+                                print_green(f'Compliance Test-{self.test_count} for {yaml_file} successful.')
+                        if not tag_matched:
```

Comment:
Not sure what you need `tag_matched` for. You could basically just use an `else` block for `if tag in yaml_data["tags"]`

Reply:
We are validating against multiple user-provided tags. Using if-else would give result for only the first tag. So, I am using the `tag_matched` to check if no tags were matched.

Reply:
Right. My bad. I still think this could be made simpler though.

How about putting the tag matcher in a separate method `.tag_matcher()` that takes a set of tags (from `yaml_data["tags"]) as input and returns a `bool`? It's a very well defined problem and would benefit from encapsulation. If we change the tag matching logic later for some reason, the behavior of the .`run_jobs()` method (and it's unit tests) wouldn't be affected.